compiler/luci/pass/src/RemoveRedundantTranspose.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/RemoveRedundantTransposePass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/CircleOpcode.h>
+
+namespace luci
+{
+
+/// @brief Return true if first_prem[second_prem[i]] == i
+bool check_perm(const luci::CircleConst *first_perm, const luci::CircleConst *second_perm)
+{
+  assert(first_perm->rank() == 1);
+  assert(second_perm->rank() == 1);
+  assert(second_perm->size<loco::DataType::S32>() == first_perm->size<loco::DataType::S32>());
+  for (int32_t i = 0; i < static_cast<int32_t>(first_perm->size<loco::DataType::S32>()); i++)
+  {
+    if (first_perm->at<loco::DataType::S32>(second_perm->at<loco::DataType::S32>(i)) != i)
+      return false;
+  }
+  return true;
+}
+
+bool remove_consecutive_transpose_function(luci::CircleNode *node)
```

Comment:
Implementation of this function have several problems for me :(

1. Checking the type of `CircleNode`
    - There are two ways, one is using `node->opcode()` and the other one is `dynamic_cast<luci::Circlexxxx>(node)`
    - Usually, the latter one is used but this function is not consistent
1. Using `node->arg(x)`
    - Some of codes use `node->arg(x)`
    - This is technically not wrong, but readability is really bad
    - For example, if `node` is definitely `CircleTranspose`. then following code can be used
    ```cpp
    auto circle_transpose = dynamic_cast<luci::CircleTranspose *>(node);
    if (circle_transpose == nullptr)
        return false;

    auto pred_node = circle_transpose->a();
    ...
    ```
1. Using `drop()` function
    - You don't need to call this function because dead nodes will be removed automatically.
1. (optional) Applying variable name to comment
    - Line 91 has nice diagrams.
    - What is `pred_node` and what is `main_node` in that diagram? If this is applied in the diagram, it would be wonderful.

Reply:
@llFreetimell I have a little question about `drop()`. as said above, I don't have to call `drop()` in all cases?

Reply:
Yes :)
`logo::RemoveDeadNodeWithQueryPass` will automatically remove them

Reply:
Currently, `RemoveDeadNodeWithQueryPass` is called after all passes are executed. So, if `node->drop()` is not called here, the connection between the `node` and the predecessor will be alive until `RemoveDeadNodeWithQueryPass`. This can cause unintended results in the passes between this pass and `RemoveDeadNodeWithQueryPass`.

To prevent such cases, how about running `RemoveDeadNodeWithQueryPass` after each pass (I think type/shape inference can be called after each pass as well)?

Reply:
@jinevening `replace` changes the user of output, which means the dropped nodes are excluded from the list of `active_node`. Since all passes starts from the `active_node` list, there is no problem I think.

Reply:
@mhs4670go `replace` changes the connections with successors. `drop` removes the connections with predecessors.

If `drop()` is not called, the connections with predecessors are alive. So, the node (dead node) is detected when `loco::succs` is called.

Reply:
@jinevening This is what I understand about it.
When the node changes its input like `at(0)->node(node)`, `node` method changes its `Use` list.
Say there is a node having two input and one output.
```
[input1] [input2]
      \      /
    [the_node]        [node_to_replace]
          |
      [output]
```
If we replace `the_node` with `node_to_replace`
```
                            [input1]     [input2] 
                               \          /
    [the_node]              [node_to_replace]
                                    |
                                 [output]
```
- the list of `active_node` : `output`, `node_to_replace`, `input1`, `input2`
- `loco::succs` of `node_to_replace` : `output`
- `loco::pred` of `node_to_replace` : `input1`, `input2`

Reply:
@mhs4670go  The problem is `loco::succs` of `input1` or `input2`.

`loco::succs` of `input1` is not only `node_to_replace` but also `the_node` (dead node). `the_node->drop()` removes the connections between `the_node` and its predecessors (`input1`, `input2`), preventing such cases.

Reply:
@jinevening But the point is why we should remove the connection. Eventually `the_node` will be removed at last.
If some optimizations for `the_node` was done, those will be removed too because all the nodes after `the_node` is dead.

Reply:
@jinevening @llFreetimell @mhs4670go while find node to optimize on pass, it use a `loco::active_nodes(loco::output_nodes(Graph *g))` to find a target node to optimize. and this return all nodes from `loco::output_nodes(Graph *g)` to it's predecessors. so, if one of active node replaced on one path, another node cannot see this node as a target and only successor case will be shown. so. I think just go without 'drop()' would be okay for it. Thanks for all! if any Idea to raise on, please comment galdly!